aten/src/ATen/native/GridSampler.cpp

```diff
@@ -0,0 +1,780 @@
+            iy_sw_cl = clip_coordinates(iy_sw, inp_H);
+            ix_se_cl = clip_coordinates(ix_se, inp_W);
+            iy_se_cl = clip_coordinates(iy_se, inp_H);
+          } else {
+            ix_nw_cl = ix_nw;
+            iy_nw_cl = iy_nw;
+            ix_ne_cl = ix_ne;
+            iy_ne_cl = iy_ne;
+            ix_sw_cl = ix_sw;
+            iy_sw_cl = iy_sw;
+            ix_se_cl = ix_se;
+            iy_se_cl = iy_se;
+          }
+
+          scalar_t gix = static_cast<scalar_t>(0), giy = static_cast<scalar_t>(0);
+          scalar_t *gOut_ptr_NCHW = gOut_ptr + n * gOut_sN + h * gOut_sH + w * gOut_sW;
+          scalar_t *gInp_ptr_NC = gInp_ptr + n * gInp_sN;
+          scalar_t *inp_ptr_NC = inp_ptr_N;
+          // calculate bilinear weighted pixel value and set output pixel
+          for (int c = 0; c < C; ++c, gOut_ptr_NCHW += gOut_sC, gInp_ptr_NC += gInp_sC, inp_ptr_NC += inp_sC) {
+            scalar_t gOut = *gOut_ptr_NCHW;
+
+            // calculate and set grad_input
+            safe_add_2d(gInp_ptr_NC, iy_nw_cl, ix_nw_cl, gInp_sH, gInp_sW, inp_H, inp_W, nw * gOut);
+            safe_add_2d(gInp_ptr_NC, iy_ne_cl, ix_ne_cl, gInp_sH, gInp_sW, inp_H, inp_W, ne * gOut);
+            safe_add_2d(gInp_ptr_NC, iy_sw_cl, ix_sw_cl, gInp_sH, gInp_sW, inp_H, inp_W, sw * gOut);
+            safe_add_2d(gInp_ptr_NC, iy_se_cl, ix_se_cl, gInp_sH, gInp_sW, inp_H, inp_W, se * gOut);
+
+            // calculate grad_grid
+            if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_nw_cl, ix_nw_cl, inp_H, inp_W)) {
+              scalar_t nw_val = inp_ptr_NC[iy_nw_cl * inp_sH + ix_nw_cl * inp_sW];
+              gix -= nw_val * (iy_se - iy) * gOut;
+              giy -= nw_val * (ix_se - ix) * gOut;
+            }
+            if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_ne_cl, ix_ne_cl, inp_H, inp_W)) {
+              scalar_t ne_val = inp_ptr_NC[iy_ne_cl * inp_sH + ix_ne_cl * inp_sW];
+              gix += ne_val * (iy_sw - iy) * gOut;
+              giy -= ne_val * (ix - ix_sw) * gOut;
+            }
+            if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_sw_cl, ix_sw_cl, inp_H, inp_W)) {
+              scalar_t sw_val = inp_ptr_NC[iy_sw_cl * inp_sH + ix_sw_cl * inp_sW];
+              gix -= sw_val * (iy - iy_ne) * gOut;
+              giy += sw_val * (ix_ne - ix) * gOut;
+            }
+            if (padding_mode != GridSamplerPadding::Zeros || within_bounds_2d(iy_se_cl, ix_se_cl, inp_H, inp_W)) {
+              scalar_t se_val = inp_ptr_NC[iy_se_cl * inp_sH + ix_se_cl * inp_sW];
+              gix += se_val * (iy - iy_nw) * gOut;
+              giy += se_val * (ix - ix_nw) * gOut;
+            }
+          }
+
+          // un-normalize grad_grid values back to [-1, 1] constraints
+          gix = gix * (inp_W - 1) / 2;
+          giy = giy * (inp_H - 1) / 2;
+
+          // assuming grad_grid is contiguous
+          gGrid_ptr_NHW[0] = gix;
```

Comment:
THNN version does an addition,  but each `grad_grid` index is only processed once. So assign is more efficient.